catalog/dags/database/catalog_cleaner/catalog_cleaner.py

```diff
@@ -0,0 +1,216 @@
+"""
+Catalog Data Cleaner DAG
+
+Use the TSV files created during the cleaning step of the ingestion process to bring
+the changes into the catalog database and make the updates permanent.
+
+The DAG has a structure similar to the batched_update DAG, but with a few key
+differences:
+  1. Given the structure of the TSV, it updates a single column at a time.
+  2. The batch updates are parallelized to speed up the process. The maximum number of
+     active tasks is limited to 3 (at first to try it out and) to avoid overwhelming
+     the database.
+  3. It needs slightly different SQL queries to update the data. One change for example,
+     is that it only works with the `image` table given that is the only one where the
+     cleaning steps are applied to in the ingestion server.
+"""
+
+import logging
+from datetime import timedelta
+
+from airflow.decorators import dag, task
+from airflow.models import Variable
+from airflow.models.abstractoperator import AbstractOperator
+from airflow.models.param import Param
+from airflow.operators.python import get_current_context
+from airflow.utils.trigger_rule import TriggerRule
+
+from common import slack
+from common.constants import DAG_DEFAULT_ARGS, POSTGRES_CONN_ID
+from common.sql import (
+    RETURN_ROW_COUNT,
+    PGExecuteQueryOperator,
+    PostgresHook,
+    single_value,
+)
+from database.batched_update.batched_update import run_sql
+from database.catalog_cleaner import constants
+
+
+logger = logging.getLogger(__name__)
+
+
+@task
+def count_dirty_rows(temp_table_name: str, task: AbstractOperator = None):
+    """Get the number of rows in the temp table before the updates."""
+    count = run_sql.function(
+        dry_run=False,
+        sql_template=f"SELECT COUNT(*) FROM {temp_table_name}",
+        query_id=f"{temp_table_name}_count",
+        handler=single_value,
+        task=task,
+    )
+    logger.info(f"Found {count:,} rows in the `{temp_table_name}` table.")
+    return count
+
+
+@task
+def get_batches(total_row_count: int, batch_size: int) -> list[tuple[int, int]]:
+    """Return a list of tuples with the start and end row_id for each batch."""
+    return [(i, i + batch_size) for i in range(0, total_row_count, batch_size)]
+
+
+@task(map_index_template="{{ index_template }}")
+def update_batch(
+    batch: tuple[int, int],
+    temp_table_name: str,
+    column: str,
+    task: AbstractOperator = None,
+):
+    batch_start, batch_end = batch
+    logger.info(f"Going through row_id {batch_start:,} to {batch_end:,}.")
+
+    # Includes the formatted batch range in the context to be used as the index
+    # template for easier identification of the tasks in the UI.
+    context = get_current_context()
+    context["index_template"] = f"{batch_start}__{batch_end}"
```

Comment:
Woah, this is neat!! I didn't know this could be evaluated _after_ the task had started 😮

Reply:
Same, TIL!!